Import existing groups

ccrockatt opened the pull request:
* Added a flag to specify whether to import groups that already exist in bitbucket along with functionality to import those groups into the terraform state

* Changed the error checking when importing an existing group for bitbucket so that the duplicate group error is not thrown
* Added two acceptance tests
* Updated the ReadMe.md with accurate instructions since the previous ones were not working for us
* Updated the Makefile to set the maximum number of files allowed to be open to 1024 since 256 was too few.  We called `ulimit -n 1024` to do this for the testacc and testacc-bitbucket targets.
* Changed the start-docker-compose.sh to use the bitbucket server environment variable so that it doesn't always refer to localhost
* Skipped some testcases that are consistently failing in master branch
* Added a version.env file to be used to control the versioning from CICD builds
* Set version to 1.4.0

gavinbunney commented on bitbucket/data_plugin_test.go:
```diff
@@ -7,6 +7,7 @@ import (
 )
 
 func TestAccBitbucketDataPlugin_notifyer(t *testing.T) {
+	t.Skip("Skipping testing in CI environment")
```
These [pass ok in the CI environment](https://travis-ci.org/github/gavinbunney/terraform-provider-bitbucketserver/builds/615187581), is there a reason to skip all those tests?

eli-guidewire replied:
Hi @gavinbunney, we can remove this change to skip the tests, but FYI, the tests are really flaky as seen in this recent [build](https://travis-ci.org/github/gavinbunney/terraform-provider-bitbucketserver/builds/671483951) as well.

Kamahl19 replied:
I only observed `TestAccBitbucketPluginConfig` to be flaky while you are skipping 3 different tests here. Are you sure all of them are flaky? 

I personally dont believe in skipping the tests. We should rather debug this and make them stable.

ccrockatt replied:
We are running all of the tests on our CI system as well and this test fails sometimes due to the plugin not being loaded and then the destruction of the resource fails with an API Error 500. This is likely due to the bitbucket status page replying with 200 response codes which the tests rely on to confirm that the Bitbucket server is up before the server is fully available. As a result the plugin is not able to load.
If the tests are run after a longer time (giving Bitbucket time to fully start up) then this test and the other plugin tests will pass consistently.
We can add these tests back for now and figure out how to resolve these failures later.

ccrockatt replied:
@kamahl19 that is our preference as well but we don't always have enough time to both fix existing problems and implement new functionality.

ccrockatt replied:
We've included these tests again

gavinbunney commented on bitbucket/resource_group.go:
```diff
@@ -21,6 +24,11 @@ func resourceGroup() *schema.Resource {
 				Required: true,
 				ForceNew: true,
 			},
+			"import_if_exists": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
```
This is a bit of a terraform anti-pattern. The normal process is to do an explicit import of a group outside of the plan/apply phase. What is the workflow you're trying to fix with this?

eli-guidewire replied:
We realize the anti-pattern, hence a flag is created with the default value of false. 

Our use case for the import_if_exists flag is for migrations where the groups would already exist in the RDS snapshots that we would restore from. In this case, we can't import the state because we haven't run the terraform apply yet to bring up the bitbucket server.

gavinbunney replied:
hmm ok that is a bit odd and i'm ok for it to go through as it's a little benign - can you also update the docs here: https://github.com/gavinbunney/terraform-provider-bitbucketserver/blob/master/docusaurus/docs/resource_user.md

melvilgit replied:
Updated the resourc_group.md file doc for  add import_if_exist functionality.

gavinbunney commented on scripts/start-docker-compose.sh:
```diff
@@ -14,4 +14,4 @@ echo "--> Starting docker-compose"
 
 echo "--> Wait for bitbucket to be ready"
-bash ${DIR}/wait-for-url.sh --url http://localhost:7990/status --timeout 600
+bash ${DIR}/wait-for-url.sh --url ${BITBUCKET_SERVER}/status --timeout 600
```
This is failing on CI: https://travis-ci.org/github/gavinbunney/terraform-provider-bitbucketserver/builds/706988452#L228

gavinbunney commented on version.env:
```diff
@@ -0,0 +1,10 @@
+#We use semantic versioning:
+#version x.y.z
+#x is an api version and if breaking api changes are introduced this should change, otherwise it remains constant
+#y is a new feature version and this only changes if new features are changed but bugs are not fixed
+#z is a patch version that changes if a bug is fixed for a previous feature version.
+# See https://semver.org/ for more information
+
+#To use this file call `source version.env` from a bash terminal
+
+export BITBUCKET_PROVIDER_VERSION=1.4.0
```
Version is controlled by the build number/tag in Travis, please remove

ccrockatt replied:
How can we retrieve the version number for our forked repositories when we are building and testing the provisioner, is there a file that Travis uses? If not it would be nice to have a way to synchronize all builds of this project.  This idea has been borrowed from maven dependency versioning which has stood the test of time.

gavinbunney replied:
The travis job will publish a new release against the releases page for each tag; so the versioning is not required until it's actually merged to master and included in a release. + we are in golang-land not maven hah

You can use a script [similar to this](https://gavinbunney.github.io/terraform-provider-kubectl/docs/provider.html#install-latest-version) to fetch the latest version available

melvilgit replied:
We have removed  version.env and updated the PR.

ccrockatt replied:
@gavinbunney can you explain in more detail how you're recommending the version be retrieved? Thanks

ccrockatt replied:
I'm currently using `curl -Ls https://api.github.com/repos/gavinbunney/terraform-provider-bitbucketserver/releases/latest | jq -r ".tag_name"`

gavinbunney replied:
This script fetches and installs the latest version: https://gavinbunney.github.io/terraform-provider-bitbucketserver/docs/provider.html#install-latest-version

Kamahl19 commented:
I believe this PR should be split to multiple smaller PRs each taking care of it's own thing. It's never good to merge stuff like this, its hard to reason about multiple different changes and hard to revert them if one of them is problematic

Kamahl19 commented:
I have updated README.md in this PR which includes some of your changes and adds others to make development and testing instructions clear and simple https://github.com/gavinbunney/terraform-provider-bitbucketserver/pull/13

Kamahl19 commented on Makefile:
```diff
@@ -20,11 +20,15 @@ test: fmtcheck
 
 testacc: fmtcheck
-	TF_ACC=1 go test $(TEST) -v $(TESTARGS) -timeout 120m -count=1
+	#The ulimit command is required to allow the tests to open more than the default 256 files as set on MacOS. The tests will fail without this. It must be done as one
```
Acceptance tests are running fine for me on OSX locally. How can I reproduce the problem this is fixing?

ccrockatt replied:
@Kamahl19 what is the value when you type in "ulimit -n", we see for our Mac OS computers that the value is 256 which is apparently too low.

Kamahl19 replied:
its 256 as well

ccrockatt replied:
Interesting.  When we run without the ulimit change we see the following output for many or maybe all of the tests
```=== RUN   TestAccBitbucketResourceRepositoryPermissionsUser
--- FAIL: TestAccBitbucketResourceRepositoryPermissionsUser (0.00s)
    testing.go:568: Step 0 error: Error creating temporary file for config: open /var/folders/76/nb0x15hd77s7hvrrsw_z2cjm0000gp/T/tf-test084352621/main.tf: too many open files

ccrockatt commented:
> I believe this PR should be split to multiple smaller PRs each taking care of it's own thing. It's never good to merge stuff like this, its hard to reason about multiple different changes and hard to revert them if one of them is problematic

We included all the changes that we made in order to get the functionality working.

ccrockatt commented:
> I have updated README.md in this PR which includes some of your changes and adds others to make development and testing instructions clear and simple #13

We've reverted the changes to the Readme.md

Kamahl19 commented on bitbucket/resource_group.go:
```diff
@@ -61,6 +79,12 @@ func resourceGroupRead(d *schema.ResourceData, m interface{}) error {
 	return fmt.Errorf("unable to find a matching group %s", groupName)
 }
 
+func resourceGroupUpdate(d *schema.ResourceData, m interface{}) error {
```
why even adding the update method? It's not required

eli-guidewire replied:
As mentioned in the inline comments:

> The only attribute in the schema that does not have "ForceNew: true" is "import_if_exists",
> so we are not actually updating any groups in Bitbucket, we just need to read and return.

When the forceNew is not set for import_if_exists, we can the following error without the update function implemented. This is because there's no API to update the group attribute. 
```
1 error occurred:
                * resource bitbucketserver_group: No Update defined, must set ForceNew on: []string{"import_if_exists"}
```

Kamahl19 commented on bitbucket/resource_group.go:
```diff
@@ -29,12 +37,22 @@ func resourceGroupCreate(d *schema.ResourceData, m interface{}) error {
 	_, err := client.Post(fmt.Sprintf("/rest/api/1.0/admin/groups?name=%s", url.QueryEscape(groupName)), nil)
 	if err != nil {
-		return err
+		if importIfExists && strings.Contains(err.Error(), "API Error: 409") {
```
can you check the API status code instead of string?

eli-guidewire replied:
We've updated this to use the API status code.

eli-guidewire commented:
@gavinbunney Can you please take another look and see if we're ready to merge the PR? Please let us know if you have more questions or concerns. Thanks.
